### PR TITLE
Gabeaws/jenkins/complete example

### PR DESCRIPTION
**Pull request by gabebatista**
**Issue number:** 176

## Summary

Adds simple, but complete, example of deploying Jenkins on its own.

### Changes

Adds simple, but complete, example of deploying Jenkins on its own.
Fixes Jenkins Route53 record name to properly align with certificate

### User experience

Users now have a simple example demonstrating how to deploy Jenkins on its own

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

<details>
<summary>Is this a breaking change?</summary>
No
</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created might not be successful.

### CODE REVIEW

File: modules/jenkins/examples/complete/dns.tf [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+data "aws_route53_zone" "root" {
+  name         = local.fully_qualified_domain_name
+  private_zone = false
+}
+
+# resource "aws_route53_zone" "root" {
+#   name = local.fully_qualified_domain_name
+#   #checkov:skip=CKV2_AWS_38: DNSSEC signing disabled by design
+#   #checkov:skip=CKV2_AWS_39: Query logging disabled by design
+# }
```
Comment (henrykie) [inline modules/jenkins/examples/complete/dns.tf:13]: Lets remove this

File: modules/jenkins/examples/complete/dns.tf [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    evaluate_target_health = true
+  }
+}
+
+##########################################
+# Jenkins Certificate Management
+##########################################
+
+resource "aws_acm_certificate" "jenkins" {
+  domain_name       = "jenkins.${data.aws_route53_zone.root.name}"
```
Comment (henrykie) [inline modules/jenkins/examples/complete/dns.tf:31]: lets reference the `aws_route53_record.jenkins.name` field for this instead

File: modules/jenkins/examples/complete/dns.tf [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  allow_overwrite = true
+  name            = each.value.name
+  records         = [each.value.record]
+  ttl             = 60
+  type            = each.value.type
+  zone_id         = data.aws_route53_zone.root.id
+}
+
+# tflint-ignore: terraform_required_providers
```
Comment (henrykie) [inline modules/jenkins/examples/complete/dns.tf:60]: Whats up with this? We should include a versions.tf or providers.tf file inside the sample that specifies which versions of the provider are used. dependabot will keep these up to date.

File: modules/jenkins/examples/complete/local.tf [hunk 4]
```diff
@@ -0,0 +1,43 @@
+data "aws_availability_zones" "available" {}
+
+
+locals {
+  # Simple Build Pipeline Configuration
+  fully_qualified_domain_name = "example.com" # update this to the root domain name you own
```
Comment (henrykie) [inline modules/jenkins/examples/complete/local.tf:6]: Lets shift these two variables please. fully_qualified_domain_name should be a variable that is injected at apply time.

File: modules/jenkins/examples/complete/local.tf [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  # IPV4 CIDR blocks that need Jenkins access. For example:
+  # "<my personal ip>/32"
+
+  # tflint-ignore: terraform_unused_declarations
+  allowlist = []
```
Comment (henrykie) [inline modules/jenkins/examples/complete/local.tf:12]: remove

Comment (henrykie) [reply modules/jenkins/examples/complete/local.tf:12]: If tflint is telling us its unused, then we can remove it

File: modules/jenkins/examples/complete/local.tf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  # VPC Configuration
+  vpc_cidr_block       = "10.0.0.0/16"
+  public_subnet_cidrs  = ["10.0.1.0/24", "10.0.2.0/24"]
+  private_subnet_cidrs = ["10.0.3.0/24", "10.0.4.0/24"]
+
+  tags = {
+    environment = "cgd"
+  }
+  azs = slice(data.aws_availability_zones.available.names, 0, 2)
+}
```
Comment (henrykie) [inline modules/jenkins/examples/complete/local.tf:43]: Keep all of this, but wipe the other locals. Move fully_qualified_domain_name to a variable injected at runtime. That way we can actually automate deployments into a hosted zone we've created ahead of time in the CI account.

File: modules/jenkins/examples/complete/main.tf [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# tflint-ignore: terraform_required_version
```
Comment (henrykie) [inline modules/jenkins/examples/complete/main.tf:1]: Same thing here. Its telling you to specify which versions of the provider you are using. That way we don't have samples / examples on multiple versions of the provider.

File: modules/jenkins/examples/complete/vpc.tf [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+resource "aws_vpc" "build_pipeline_vpc" {
+  cidr_block = local.vpc_cidr_block
+  tags = merge(local.tags,
+    {
+      Name = "build-pipeline-vpc"
+    }
+  )
+  enable_dns_hostnames = true
+  #checkov:skip=CKV2_AWS_11: VPC flow logging disabled by design
+}
```
[the remark this answers is not in the document]
Comment (henrykie) [reply modules/jenkins/examples/complete/vpc.tf:14]: Its funny to me that its still alarming on this... When you run the precommit checks does this flag? Or is it only on the GH side?

File: modules/jenkins/examples/complete/vpc.tf [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+# Set default SG to restrict all traffic
+resource "aws_default_security_group" "default" {
+  vpc_id = aws_vpc.build_pipeline_vpc.id
```
Comment (henrykie) [inline modules/jenkins/examples/complete/vpc.tf:18]: Can we rename the resources that are `build_pipeline` specific to things that are more focused on this "jenkins" specific example? Its the `jenkins_vpc` in this context.

File: modules/jenkins/examples/complete/vpc.tf [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+##########################################
+# Internet Gateway
+##########################################
+
+resource "aws_internet_gateway" "igw" {
+  vpc_id = aws_vpc.build_pipeline_vpc.id
+  tags = merge(local.tags,
+    {
+      Name = "build-pipeline-igw"
```
Comment (henrykie) [inline modules/jenkins/examples/complete/vpc.tf:59]: Same thing as above - naming.

File: modules/jenkins/examples/complete/vpc.tf [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+  # public route to the internet
+  route {
+    cidr_block = "0.0.0.0/0"
+    gateway_id = aws_internet_gateway.igw.id
+  }
+
+  tags = merge(local.tags,
+    {
+      Name = "build-pipeline-public-rt"
```
Comment (henrykie) [inline modules/jenkins/examples/complete/vpc.tf:79]: naming

File: modules/jenkins/examples/complete/vpc.tf [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  route_table_id = aws_route_table.public_rt.id
+  subnet_id      = aws_subnet.public_subnets[count.index].id
+}
+
+resource "aws_eip" "nat_gateway_eip" {
+  depends_on = [aws_internet_gateway.igw]
+  #checkov:skip=CKV2_AWS_19:EIP associated with NAT Gateway through association ID
+  tags = merge(local.tags,
+    {
+      Name = "build-pipeline-nat-eip"
```
Comment (henrykie) [inline modules/jenkins/examples/complete/vpc.tf:95]: naming

File: modules/jenkins/examples/complete/vpc.tf [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    }
+  )
+}
```
[the remark this answers is not in the document]
Comment (henrykie) [reply modules/jenkins/examples/complete/vpc.tf:98]: Does this flag on precommit?

File: modules/jenkins/examples/complete/vpc.tf [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+  # route to the internet through NAT gateway
+  route {
+    cidr_block     = "0.0.0.0/0"
+    nat_gateway_id = aws_nat_gateway.nat_gateway.id
+  }
+
+  tags = merge(local.tags,
+    {
+      Name = "build-pipeline-private-rt"
```
Comment (henrykie) [inline modules/jenkins/examples/complete/vpc.tf:111]: naming

File: modules/jenkins/examples/complete/vpc.tf [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+  subnet_id      = aws_subnet.private_subnets[count.index].id
+}
+
+# tflint-ignore: terraform_required_providers
+resource "aws_nat_gateway" "nat_gateway" {
+  allocation_id = aws_eip.nat_gateway_eip.id
+  subnet_id     = aws_subnet.public_subnets[0].id
+  tags = merge(local.tags,
+    {
+      Name = "build-pipeline-nat"
```
Comment (henrykie) [inline modules/jenkins/examples/complete/vpc.tf:128]: naming

**Review by henrykie — CHANGES_REQUESTED** [review]
Couple of comments to resolve, but overall it looks good.

Make these changes, and then commit these updates with the following message:

feat(jenkins example): Adding simple example deployment of Jenkins for automated testing

File: modules/jenkins/examples/complete/vpc.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (gabebatista) [reply modules/jenkins/examples/complete/vpc.tf:14]: I didn't see in the precommit, so it's GH only

File: modules/jenkins/examples/complete/vpc.tf (hunk 13 above)
[the remark this answers is not in the document]
Comment (gabebatista) [reply modules/jenkins/examples/complete/vpc.tf:98]: This also didnt come up as an issue on the precommit.

File: modules/jenkins/examples/complete/dns.tf (hunk 2 above)
Comment (gabebatista) [reply modules/jenkins/examples/complete/dns.tf:31]: This change causes a `Cycle` error.
Screenshot 2024-09-12 at 4 19 01 PM

**Review by henrykie — DISMISSED** [review]

File: modules/jenkins/examples/complete/versions.tf [hunk 16]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.0"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "5.62.0"
```
Comment (jorisdon) [inline modules/jenkins/examples/complete/versions.tf:7]: This needs to match the version in `modules/jenkins/versions.tf` or else `terraform init` fails with:
```
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider hashicorp/aws: no available releases match the given constraints 5.62.0, 5.66.0
```

File: modules/jenkins/examples/complete/variables.tf [hunk 17]
```diff
@@ -0,0 +1,10 @@
+variable "fully_qualified_domain_name" {
+  type        = string
+  description = "The fully qualified domain name (FQDN) to be used for jenkins"
+}
+
+variable "jenkins_agent_secret_arns" {
+  type        = list(string)
+  description = "A list of secretmanager ARNs (wildcards allowed) that contain any secrets which need to be accessed by the Jenkins service."
+  default     = null
```
Comment (jorisdon) [inline modules/jenkins/examples/complete/variables.tf:9]: Probably a good idea to make the user explicitly provide this, or provide a better default.
Currently, this default causes an error:
```
│ Error: Invalid function argument
│
│   on ..\..\iam.tf line 75, in data "aws_iam_policy_document" "jenkins_default_policy":
│   75:     for_each = length(var.jenkins_agent_secret_arns) > 0 ? [1] : []
│     ├────────────────
│     │ while calling length(value)
│     │ var.jenkins_agent_secret_arns is null
```

File: modules/jenkins/examples/complete/variables.tf [hunk 18]
```diff
@@ -0,0 +1,10 @@
+variable "fully_qualified_domain_name" {
+  type        = string
+  description = "The fully qualified domain name (FQDN) to be used for jenkins"
```
Comment (jorisdon) [inline modules/jenkins/examples/complete/variables.tf:3]: The FQDN here should NOT be prefixed with "jenkins." correct? As the sample takes care of that elsewhere? If so, might be a good idea to clarify this in the description.

File: modules/jenkins/examples/complete/variables.tf (hunk 17 above)
Comment (jorisdon) [reply modules/jenkins/examples/complete/variables.tf:9]: Defaulting this to `[]` works

**Review by jorisdon — CHANGES_REQUESTED** [review]

**Review by jorisdon — DISMISSED** [review]

**Review by henrykie — APPROVED** [review]

**Pull request merged**